src/gui/emojimodel.cpp

```diff
@@ -0,0 +1,1577 @@
+/*
+ * Copyright (C) by Felix Weilbach <felix.weilbach@nextcloud.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include <QByteArray>
+#include <QDebug>
+#include <QtGlobal>
+
+#include "emojimodel.h"
+
+namespace OCC {
+
+QObject *EmojiModel::singletonProvider(QQmlEngine *qmlEngine, QJSEngine *jsEngine)
+{
+    Q_UNUSED(qmlEngine)
+    Q_UNUSED(jsEngine)
+
+    static auto instance = new EmojiModel;
+    return instance;
+}
+
+QVariantList EmojiModel::history()
+{
+    return _settings.value("Editor/emojis", QVariantList()).toList();
+}
+
+QVariantList EmojiModel::filterModel(const QString &filter)
+{
+    QVariantList result;
+
+    for (const QVariant &e : people) {
```

Comment:
Duplicate code here 😀 maybe try to refactor into a method?

Only a wild idea. But is there any way to use the OSes native Emoji picker? Like the one on Windows looks quite nice and would feel very nice. I think on Linuxes there is one too?

Reply:
> Duplicate code here grinning maybe try to refactor into a method?

Yes, good idea :)

> Only a wild idea. But is there any way to use the OSes native Emoji picker? Like the one on Windows looks quite nice and would feel very nice. I think on Linuxes there is one too?

That would be great and I thought about it, but then we need to deal with special code for Windows, Linux (for Linux there is a difference if you use KDE or GNOME or something else) and macOS. That would defeat the purpose of using Qt, but I think in the long term Qt should provide an emoji picker that may be able to use the one provided by the system. Because some desktop environments do not provide an emoji picker we will need a fallback like this in any case.
Maybe we can also just merge this generic emoji picker and someone (maybe you ;) ) creates later a pull request that will open the system emoji picker if it is there.
What do you think about this? @mgallien @camilasan @allexzander


> Otherwise you will have to keep up with the features of the native ones like skin tone e.g. or all the new emojis coming every now and then.

I think it's a compromise at the moment. It's not perfect but better than nothing.